File: strands-ts/src/conversation-manager/pin-message.ts
import { Message, type ToolUseBlock, type ToolResultBlock } from '../types/messages.js'

/**
 * Check if a message is pinned, including tool-pair partner protection.
 * Returns `true` if the message at `index` is pinned, or if it is the
 * adjacent tool-pair partner (toolUse/toolResult) of a pinned message,
 * matched by toolUseId.
 *
 * @param messages - The full messages array.
 * @param index - The index to check.
 * @returns `true` if the message or its tool-pair partner is pinned.
 */
export function isPinned(messages: Message[], index: number): boolean {
  const msg = messages[index]!

  if (msg.metadata?.custom?.pinned === true) return true

  // Tool-pair partner protection: if this is a toolResult, check if prev toolUse is pinned
  const toolResultBlocks = msg.content.filter((b): b is ToolResultBlock => b.type === 'toolResultBlock')
  if (toolResultBlocks.length > 0 && index > 0) {
    const prev = messages[index - 1]!
    if (prev.metadata?.custom?.pinned === true) {
      const resultIds = new Set(toolResultBlocks.map((b) => b.toolUseId))
      if (prev.content.some((b) => b.type === 'toolUseBlock' && resultIds.has((b as ToolUseBlock).toolUseId))) {
        return true
      }
    }
  }

  // Tool-pair partner protection: if this is a toolUse, check if next toolResult is pinned
  const toolUseBlocks = msg.content.filter((b): b is ToolUseBlock => b.type === 'toolUseBlock')
  if (toolUseBlocks.length > 0 && index + 1 < messages.length) {
    const next = messages[index + 1]!
    if (next.metadata?.custom?.pinned === true) {
      const useIds = new Set(toolUseBlocks.map((b) => b.toolUseId))
      if (next.content.some((b) => b.type === 'toolResultBlock' && useIds.has((b as ToolResultBlock).toolUseId))) {
        return true
      }
    }
  }

  return false
}

/**
 * Pin a message so it is protected from eviction during context reduction.
 * Mutates the message in place by setting `metadata.custom.pinned = true`.
 *
 * @param messages - The messages array containing the message to pin.
 * @param index - The index of the message to pin.
 */
export function pinMessage(messages: Message[], index: number): void {
  const message = messages[index]!
  message.metadata = {
    ...message.metadata,
    custom: { ...message.metadata?.custom, pinned: true },
  }
}

/**
 * Pin the first N messages in the array permanently.
 *
 * @param messages - The messages array.
 * @param count - Number of messages from the start to pin.
 */
export function applyPinFirst(messages: Message[], count: number): void {
  for (let i = 0; i < Math.min(count, messages.length); i++) {
    pinMessage(messages, i)
  }
}

/**
 * Partition a range of messages into pinned (protected) and unpinned arrays.
 *
 * @param messages - The full messages array.
 * @param start - Start index of the range (inclusive).
 * @param end - End index of the range (exclusive).
 * @returns A tuple of [pinned, unpinned] message arrays.
 */
export function partitionPinned(messages: Message[], start: number, end: number): [Message[], Message[]] {
  const pinned: Message[] = []
  const unpinned: Message[] = []
  for (let i = start; i < end; i++) {
    if (isPinned(messages, i)) {
      pinned.push(messages[i]!)
    } else {
      unpinned.push(messages[i]!)
    }
  }
  return [pinned, unpinned]
}

/**
 * Unpin a message so it can be evicted during context reduction.
 * Mutates the message in place by removing the `pinned` flag from metadata.
 *
 * @param messages - The messages array containing the message to unpin.
 * @param index - The index of the message to unpin.
 */
export function unpinMessage(messages: Message[], index: number): void {
  const message = messages[index]!
  const { pinned: _, ...restCustom } = message.metadata?.custom ?? {}
  const { custom: __, ...restMetadata } = message.metadata ?? {}
  const hasCustom = Object.keys(restCustom).length > 0
  const hasMetadata = hasCustom || Object.keys(restMetadata).length > 0
  if (hasMetadata) {
    message.metadata = { ...restMetadata, ...(hasCustom ? { custom: restCustom } : {}) }
  } else {
    delete message.metadata
  }
}

File: strands-py/src/strands/agent/conversation_manager/pin_message.py
"""Message pinning utilities for protecting messages from context eviction."""

from ...types.content import Message, Messages


def _get_tool_use_ids(message: Message) -> set[str]:
    """Extract toolUseIds from toolUse or toolResult blocks in a message."""
    ids: set[str] = set()
    for content in message.get("content", []):
        if isinstance(content, dict):
            if "toolUse" in content:
                tool_id = content["toolUse"].get("toolUseId")
                if tool_id:
                    ids.add(tool_id)
            elif "toolResult" in content:
                tool_id = content["toolResult"].get("toolUseId")
                if tool_id:
                    ids.add(tool_id)
    return ids


def _has_pinned_flag(message: Message) -> bool:
    """Check if a message has metadata.custom.pinned set to True."""
    metadata = message.get("metadata")
    return metadata is not None and metadata.get("custom", {}).get("pinned") is True


def is_pinned(messages: Messages, index: int) -> bool:
    """Check if a message is pinned, including tool-pair partner protection.

    Returns True if the message at index is pinned, or if its adjacent
    tool-pair partner (toolUse/toolResult matched by toolUseId) is pinned.

    Args:
        messages: The full messages array.
        index: The index to check.

    Returns:
        True if the message or its tool-pair partner is pinned.
    """
    if _has_pinned_flag(messages[index]):
        return True

    # Check if adjacent partner shares a toolUseId and is pinned
    my_ids = _get_tool_use_ids(messages[index])
    if not my_ids:
        return False

    for neighbor_index in (index - 1, index + 1):
        if 0 <= neighbor_index < len(messages):
            neighbor = messages[neighbor_index]
            if _has_pinned_flag(neighbor) and my_ids & _get_tool_use_ids(neighbor):
                return True

    return False



def apply_pin_first(messages: Messages, count: int) -> None:
    """Pin the first N messages in the array permanently.

    Args:
        messages: The messages array.
        count: Number of messages from the start to pin.
    """
    for i in range(min(count, len(messages))):
        pin_message(messages, i)


def partition_pinned(messages: Messages, start: int, end: int) -> tuple[list[Message], list[Message]]:
    """Partition a range of messages into pinned (protected) and unpinned arrays.

    Args:
        messages: The full messages array.
        start: Start index of the range (inclusive).
        end: End index of the range (exclusive).

    Returns:
        A tuple of (pinned, unpinned) message lists.
    """
    pinned: list[Message] = []
    unpinned: list[Message] = []
    for i in range(start, end):
        if is_pinned(messages, i):
            pinned.append(messages[i])
        else:
            unpinned.append(messages[i])
    return pinned, unpinned


def pin_message(messages: Messages, index: int) -> None:
    """Pin a message so it is protected from eviction during context reduction.

    Mutates the message in place by setting metadata.custom.pinned = True.

    Args:
        messages: The messages array.
        index: The index of the message to pin.
    """
    message = messages[index]
    metadata = message.get("metadata", {})
    custom = metadata.get("custom", {})
    custom["pinned"] = True
    metadata["custom"] = custom
    message["metadata"] = metadata


def unpin_message(messages: Messages, index: int) -> None:
    """Unpin a message so it can be evicted during context reduction.

    Mutates the message in place by removing the pinned flag from metadata.

    Args:
        messages: The messages array.
        index: The index of the message to unpin.
    """
    message = messages[index]
    metadata = message.get("metadata")
    if metadata is None:
        return

    custom = metadata.get("custom")
    if custom is None:
        return

    custom.pop("pinned", None)

    if not custom:
        del metadata["custom"]
    if not metadata:
        del message["metadata"]
